skimage/color/colorconv.py

```diff
@@ -1460,13 +1463,12 @@ def separate_stains(rgb, conv_matrix):
     from ..exposure import rescale_intensity
 
     rgb = _prepare_colorarray(rgb, force_copy=True)
-    rgb += np.e  # avoiding log artifacts
+    rgb += np.maximum(rgb, 1E-6)  # avoiding log artifacts
+    log_adjust = -np.log10(1E-6)  # used to compensate the sum above
 
-    stains = np.reshape(-np.log(rgb), (-1, 3)) @ conv_matrix
-    stains = np.reshape(stains, rgb.shape)
-    stains = rescale_intensity(stains, out_range=(0, 1))
+    stains = np.reshape(-np.log10(rgb) / log_adjust, (-1, 3)) @ conv_matrix
```

Comment:
I don't think you need this reshaping business:
```python
In [1]: np.zeros((2, 2, 3)) @ np.ones((3, 3))                                   
Out[1]: 
array([[[0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.]]])

In [2]: (np.zeros((2, 2, 3)) @ np.ones((3, 3))).shape                           
Out[2]: (2, 2, 3)
```

Reply:
You're right, results are the same. Maybe some heritage from old times? :sweat_smile: